feat: add pre-commit hooks

jackhodgkiss opened the pull request:
Add a simple playbook that can install [pre-commits](https://pre-commit.com/) and register a few simple hooks. This process is implemented as a hook that will run post control host bootstrap.

- check-yaml: perform basic yaml syntax linting
- end-of-file-fixer: identify and automatically fix missing newline
- trailing-whitespace: identify and automatically fix excessive white space
- ripsecrets: identify and prevent secrets from being committed to the branch

Note: this is currently configured as an opt-in feature and will only install and register hooks if `enable_pre_commit_hooks` is set to `true` in `install-pre-commit-hooks.yml`

jackhodgkiss committed:
feat: add initial `install-pre-commit-hooks` playbook

jackhodgkiss committed:
feat: register `pre-commit` with `git`

jackhodgkiss committed:
feat: make `install-pre-commit-hooks` opt-in

jackhodgkiss committed:
feat: add release note

priteau commented on releasenotes/notes/add-pre-commit-hooks-07ce3b82bbe1d7a3.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features:
+  - |
+    Add playbook to install `pre-commit` hooks and register them with git.
```
I think you need double backticks throughout this file.

jackhodgkiss replied:
Looking at the docs the examples use single back ticks https://docs.openstack.org/reno/latest/user/examples.html#input-file

```
fixes_command_line:
  - |
    This is a subsection. It requires setting `sections` in
```

priteau replied:
> Looking at the docs the examples use single back ticks https://docs.openstack.org/reno/latest/user/examples.html#input-file
> 
> ```
> fixes_command_line:
>   - |
>     This is a subsection. It requires setting `sections` in
> ```

I think they got it wrong… And it isn't rendered so we don't see the result. Single backtick produces italic text with reno.

See https://docutils.sourceforge.io/docs/user/rst/quickref.html for details.

jackhodgkiss replied:
Thanks for the information I have added the correct number of backticks.

Alex-Welsh commented on releasenotes/notes/add-pre-commit-hooks-07ce3b82bbe1d7a3.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features:
+  - |
+    Add playbook to install ``pre-commit`` hooks and register them with git.
```
Does pre-commit need to be `codeified`? It seems like just a regular word to me. I know it's also the name of the package but it works just as well as a generic description

priteau replied:
Sure, we could remove them. Maybe Jack really meant to use italic for this one?

jackhodgkiss replied:
It was only to add emphasis. Though I may use that more than I should.

Will clean up shortly.

priteau replied:
In this case single backtick was correct, sorry!

markgoddard reviewed (commented):
How are we expecting this to be used? Is it something that would be enabled for everyone in a given Kayobe config, or opted into per-person?

markgoddard commented on etc/kayobe/ansible/install-pre-commit-hooks.yml:
```diff
@@ -0,0 +1,22 @@
+            creates: "{{ lookup('ansible.builtin.env', 'KAYOBE_CONFIG_PATH') }}/../../.git/hooks/pre-commit"
+          args:
+            chdir: "{{ lookup('ansible.builtin.env', 'KAYOBE_CONFIG_PATH') }}"
+      when: enable_pre_commit_hooks | default(false)
```
I think I'd prefer that this gets enabled by creating a symlink. That way we don't see the skipped tasks if it's not.

jackhodgkiss replied:
To clarify. You want the `when` property removed so the playbook will run providing the the symlink exists within `hooks/control-host-bootstrap/post.d` which it currently is.

If so should I remove the symlink so it remains an opt-in.

I will ensure this behaviour and steps for enabling is documentated.

markgoddard replied:
Correct

markgoddard commented on etc/kayobe/ansible/install-pre-commit-hooks.yml:
```diff
@@ -0,0 +1,22 @@
+        - name: Register pre-commit hooks with git
+          ansible.builtin.command:
+            cmd: "{{ lookup('ansible.builtin.env', 'VIRTUAL_ENV') }}/bin/pre-commit install"
+            creates: "{{ lookup('ansible.builtin.env', 'KAYOBE_CONFIG_PATH') }}/../../.git/hooks/pre-commit"
```
Will this prevent pre-commit from being updated?

jackhodgkiss replied:
I don't believe so. Though I will investigate if this is the case.

jackhodgkiss replied:
This appears to install the hook within `.git` directory and the presence of `creates` will mean any changes to such [template](https://github.com/pre-commit/pre-commit/blob/a9f19f4cc07d6fc3e0410bdc37a3d99c944760a7/pre_commit/resources/hook-tmpl) will be ignored. Could drop the `creates` to avoid this potential issue.

markgoddard replied:
I'd be tempted to drop it, or apply it only when the previous task is unchanged

markgoddard reviewed (changes requested):
It would be nice to provide some simple docs on this in the contributor guide.

jackhodgkiss commented:
> How are we expecting this to be used? Is it something that would be enabled for everyone in a given Kayobe config, or opted into per-person?

I think having it is as an opt-in at least within SKC it the sensible move. If individual engineers want to enable it within their own checkout of the site config they can do and in situations where their is a single checkout shared enabling it would a group decision.

jackhodgkiss committed:
feat!: change when `pre-commit` hooks are installed

Initally `pre-commit` hooks were installed due the presense of a hook
for `control host bootstrap` and when `enable_pre_commit_hooks` was set
to `true`. However now `pre-commit` hooks are only installed when the
hook is present or the playbook is called, no requirement for an
`Ansible` conditional to be true.

jackhodgkiss committed:
fix: update release note for `pre-commit hooks`

jackhodgkiss committed:
feat: add `pre-commit` to contributor docs

Alex-Welsh reviewed (dismissed):
My one lingering concern is that false positives could break things but they can be skipped with --no-verify so I think the benefits outweigh the risks

markgoddard commented on etc/kayobe/ansible/install-pre-commit-hooks.yml:
```diff
@@ -0,0 +1,21 @@
+  hosts: localhost
+  gather_facts: false
+  vars:
+    enable_pre_commit_hooks: false
```
Unused?

markgoddard commented on doc/source/contributor/pre-commit.rst:
```diff
@@ -0,0 +1,38 @@
+Either use `kayobe playbook run` or add the playbook as a hook within Kayobe config such as
+within `control-host-bootstrap/post.d`.
+Once done you should find `pre-commit` is available within the `kayobe` virtualenv.
+
```
It's always nice to provide copy-pastable commands to avoid excessive brain usage

jackhodgkiss committed:
fix: remove unused variable

jackhodgkiss committed:
feat: run `pre-commit install` if package is installed or updated

jackhodgkiss committed:
feat: use variable to control `pre_commit` version

jackhodgkiss committed:
feat: provide commands within docs for pre-commit hooks setup

markgoddard reviewed (changes requested):
Actually perhaps this should go to antelope only?

Alex-Welsh commented:
@jackhodgkiss should we get this merged? I'm happy to rebase it on antelope for you if you're busy

markgoddard commented on etc/kayobe/ansible/install-pre-commit-hooks.yml:
```diff
@@ -0,0 +1,22 @@
+          ansible.builtin.pip:
+            name: pre-commit
+            version: "{{ pre_commit_version  }}"
+            virtualenv: "{{ lookup('ansible.builtin.env', 'VIRTUAL_ENV') }}"
```
Is there an issue here that it forces us to use a virtualenv when developing SKC? e.g. I don't necessarily have one set up when writing docs.

jackhodgkiss replied:
I have updated this to support installation without a virtualenv if one isn't present.

markgoddard commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,11 @@
+      - id: check-yaml
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+  - repo: https://github.com/sirwart/ripsecrets
+    rev: v0.1.7
+    hooks:
+      - id: ripsecrets
```
This one seems to require a rust installation, and mine wasn't new enough.
```
[INFO] Installing environment for https://github.com/sirwart/ripsecrets.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/home/mark/.cargo/bin/cargo', 'install', '--bins', '--root', '/home/mark/.cache/pre-commit/repohp1cqnt3/rustenv-system', '--path', '.')
return code: 101
stdout: (none)
stderr:
      Installing ripsecrets v0.1.7 (/home/mark/.cache/pre-commit/repohp1cqnt3)
        Updating crates.io index
    error: failed to compile `ripsecrets v0.1.7 (/home/mark/.cache/pre-commit/repohp1cqnt3)`, intermediate artifacts can be found at `/home/mark/.cache/pre-commit/repohp1cqnt3/target`
    
    Caused by:
      package `clap v4.5.16` cannot be built because it requires rustc 1.74 or newer, while the currently active rustc version is 1.70.0
      Either upgrade to rustc 1.74 or newer, or use
      cargo update -p clap@4.5.16 --precise ver
      where `ver` is the latest version of `clap` supporting rustc 1.70.0
Check the log at /home/mark/.cache/pre-commit/pre-commit.log
```

jackhodgkiss replied:
I think this is a problem if `rustc` is already installed. It might prioritise the existing installation as opposed to running the one included within `/home/mark/.cache/pre-commit/repohp1cqnt3`

jackhodgkiss replied:
If Rust toolchain is detected it will not install Rust for building of pre-commit hooks that have it as a build dependency [1].

Therefore, if an older version is installed it might not be compatible with specific hooks used.

However, ripsecrets v0.1.7 does work with rust 1.7.0 it only fails because Cargo does use the `cargo.lock` by default and requires `--locked` to be passed [2].

This [PR](https://github.com/pre-commit/pre-commit/pull/3163) proposes a simple change to pass `--locked` which I have tested and does work.

Maybe `cargo.conf` could force this [3]?

Not sure how you would like to proceed with this issue.

- [1] https://github.com/pre-commit/pre-commit/blob/main/pre_commit/languages/rust.py#L30
- [2] https://doc.rust-lang.org/cargo/commands/cargo-install.html#:~:text=The%20%2D%2Dlocked%20flag%20can,when%20the%20package%20was%20published.
- [3] https://doc.rust-lang.org/cargo/reference/config.html

jackhodgkiss committed:
feat: support install `pre-commit` without `kayobe-venv`
